cmd/libsnap-confine-private/tool.c

```diff
@@ -72,20 +72,52 @@ int sc_open_snap_update_ns(void)
 }
 
 void sc_call_snap_update_ns(int snap_update_ns_fd, const char *snap_name,
-			    struct sc_apparmor *apparmor)
+			    struct sc_apparmor *apparmor, bool is_user_profile)
 {
 	char *snap_name_copy SC_CLEANUP(sc_cleanup_string) = NULL;
 	snap_name_copy = sc_strdup(snap_name);
-	char *argv[] =
-	    { "snap-update-ns", "--from-snap-confine", snap_name_copy, NULL };
-	/* SNAPD_DEBUG=x is replaced by sc_call_snapd_tool_with_apparmor with
-	 * either SNAPD_DEBUG=0 or SNAPD_DEBUG=1, see that function for details. */
-	char *envp[] = { "SNAPD_DEBUG=x", NULL };
-	char profile[PATH_MAX] = { 0 };
-	sc_must_snprintf(profile, sizeof profile, "snap-update-ns.%s",
+
+	char aa_profile[PATH_MAX] = { 0 };
+	sc_must_snprintf(aa_profile, sizeof aa_profile, "snap-update-ns.%s",
 			 snap_name);
-	sc_call_snapd_tool_with_apparmor(snap_update_ns_fd, "snap-update-ns",
-					 apparmor, profile, argv, envp);
+
+	if (is_user_profile) {
```

Comment:
You could refactor this to:
```
char *argv[] = {
	"snap-update-ns",
	/* This tells snap-update-ns we are calling from snap-confine and locking is in place */
	"--from-snap-confine",
	snap_name_copy, NULL
};
char *envp[] =  { "SNAPD_DEBUG=x", NULL };
if (is_user_profile) {
    argv = ...;
    envp = ....;
}
sc_call_snapd_tool_with_apparmor(snap_update_ns_fd,
				 "snap-update-ns", apparmor,
				 aa_profile, argv, envp);
```

Reply:
Tried that initially and it turned out to be worse because of how scoping and memory management works.